### ISSUE TITLE
Skipped correlation gives different results than the Pernet implementation in Matlab

### ISSUE THREAD

**Issue by adamnarai** [raphaelvallat/pingouin#164]
I think this line in `correlation.py`:
```py
dis[i, :] = np.linalg.norm(B * B2[i, :] / bot[i], axis=1)
```
is functionally different compared to the Matlab implementation and the equation in Wilcox, R. (2004)

**Comment by raphaelvallat** [comment]
Hi @adamnarai,

I coded this function a long time ago, but from what I remember it is essentially an (optimized) translation of the Matlab code from the [robust correlation toolbox](https://github.com/CPernet/Robust-Correlations) by Cyril Pernet and Guillaume Rousselet. Furthermore, the output of the function is tested against the original toolbox:

https://github.com/raphaelvallat/pingouin/blob/91b184780d27f291ff9f894d1c80f74c0c20e146/pingouin/tests/test_correlation.py#L22-L25

That said, if you have any idea for a better implementation, and/or if you find inconsistent results with the robust corr toolbox then I'm all in for reworking this function.

Thanks!
Raphael

**Comment by adamnarai** [comment]
Hi @raphaelvallat,

This test yields the same result as the robust correlation toolbox, however I can create a different test data, where it's not the case. For example the data generated by replacing line 17 with the following:
```py
x[3], y[5] = 7, 2.6
```
yields `r=0.474, n_outl=0` in pingouin and `r=0.512, n_outl=2` in the robust correlation toolbox.

I would suggest replacing this line:
https://github.com/raphaelvallat/pingouin/blob/91b184780d27f291ff9f894d1c80f74c0c20e146/pingouin/correlation.py#L82

with
```py
dis[i, :] = np.linalg.norm(B.dot(B[i, :, None]) * B[i, :] / bot[i], axis=1)
```
By applying this change, I found the results consistent with the robust correlation toolbox.

Best,
Adam

**Comment by raphaelvallat** [comment]
Hi @adamnarai,

Thanks for looking into that! You were right, I have now updated the line and unit testing in this commit: ce7545df20cf787cadfe7c1bb2998c23240116ef

I will release a new version of Pingouin in the next couple of weeks.
Raphael

**Comment by adamnarai** [comment]
Hi @raphaelvallat,

I still found some differences in the number of outliers between the two implementations when applied on real data (even 5 vs. 1 outlier in one case). As an example, using the following values in the unit test (line 18):
```py
x[3], y[5] = 20, 3.5
```
I get `r=0.512, n_outl=2` in pingouin and `r=0.418, n_outl=1` in the robust correlation toolbox.
I suspect the difference is in the MCD estimator implementation, since the center values are also different but I couldn't reproduce the robust correlation toolbox results by trying different parametrization of the MinCovDet class.

Best,
Adam

**Comment by raphaelvallat** [comment]
Hi @adamnarai,

Re-opening this issue for visibility.

You're right, I just checked the output of MCD with the example you provide and Matlab returns: `center = [3.8123, 5.7939]` while in Python `MinCovDet().fit(X).location_ = [3.78007729, 5.88216227]`. As explained in https://github.com/scikit-learn/scikit-learn/issues/4682, this is because sklearn scales by **_N_** and not by **_N-1_** (the latter being arguably more adequate though).

I think we could implement the following fix

![image](https://user-images.githubusercontent.com/25703792/126051285-2b9ac0d5-0839-457b-8f3e-85cf477b117c.png)

which gives `r=0.418, n_outl=1` (although the center values are not exactly similar to Matlab)

Let me know what you think!
Raphael

**Comment by raphaelvallat** [comment]
I just pushed a commit with these changes on the develop branch if you want to try it out on real-world data: f2deb57b34f2dd7bb625ca30ccc93a2db3059bce

**Comment by adamnarai** [comment]
Hi @raphaelvallat,

I think the normalization only affects the covariance and not the location and there is something else causing this difference. I will try to look more into it if my time allows.

By using the `fast_mcd()` function directly we basically get the estimate without re-weighting, which is the same as `MinCovDet().fit(X).raw_location_=[3.81101474, 5.81656632]`
As an example, using your commit and `x[3], y[5] = 10, 3.5`, I got `r=0.512, n_outl=2` in Pingouin and `r=0.418, n_outl=1` in Matlab.

Best,
Adam

**Comment by raphaelvallat** [comment]
@adamnarai  Ah, you're absolutely right, my bad. I've just tried with the following (inspired from [sklearn](https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/covariance/_robust_covariance.py#L674-L762)):

```python
X = np.column_stack((x, y))
nrows, ncols = X.shape
center, cov, support, dist = fast_mcd(X, cov_computation_method=lambda x: np.cov(x, rowvar=False))
correction = np.median(dist) / chi2(X.shape[1]).isf(0.5)
dist /= correction
mask = dist < chi2(2).isf(0.025)
print(X[mask].mean(0))
```

But I still get the same center values as sklearn.MinCovDet, meaning that indeed the normalization has no effect on the location. I'm reverting the commit now.

I've tried playing around with different `support_fraction` (the `h` parameter in Matlab), but I can never get the same results as Matlab so my guess is the Matlab/Libra algorithm is different than scikit-learn, and I don't think there's anything we can do at this stage unfortunately.

Best,
Raphael

**Comment by adamnarai** [comment]
I agree, at this point that's the best we can realistically do.

For future reference:
The `fast_mcd()` function seems to give the same center values as the Matlab `mcdcov()` `raw.center`, however, the other outputs are different, even when using the same covariance normalization and the same `support_fraction` parameter.

Best,
Adam

**Comment by raphaelvallat** [comment]
Warning added in e56df016966f34c9d2f7cca882e84382dcff4d2d. This will be included in the next stable release of Pingouin.

Thanks again for looking into this,
Raphael

**Comment by raphaelvallat** [comment]
The warning has been included in the new stable version of Pingouin ([v0.4.0](https://github.com/raphaelvallat/pingouin/releases/tag/v0.4.0)). Please make sure to upgrade with `pip install --upgrade pingouin`.

Thanks,
Closing the issue.
